frontend/exporter/src/traits.rs

```diff
@@ -200,62 +199,90 @@ pub mod rustc {
                     })
                     .collect()
             }
+        }
 
-            #[tracing::instrument(level = "trace", skip(s))]
-            fn path_to(
-                self,
-                s: &S,
-                target: PolyTraitRef<'tcx>,
-                param_env: rustc_middle::ty::ParamEnv<'tcx>,
-            ) -> Option<Path<'tcx>> {
-                let tcx = s.base().tcx;
-                if predicate_equality(self.upcast(tcx), target.upcast(tcx), param_env, s) {
-                    return Some(vec![]);
+        #[tracing::instrument(level = "trace", skip(s, param_env))]
+        pub fn path_to<'tcx, S: UnderOwnerState<'tcx>>(
+            starting_points: Vec<AnnotatedPredicate<'tcx>>,
+            s: &S,
+            target: PolyTraitRef<'tcx>,
+            param_env: rustc_middle::ty::ParamEnv<'tcx>,
+        ) -> Option<(Path<'tcx>, AnnotatedPredicate<'tcx>)> {
+            let tcx = s.base().tcx;
+
+            /// A candidate projects `self` along a path reaching some
+            /// predicate. A candidate is selected when its predicate
+            /// is the one expected, aka `target`.
+            #[derive(Debug)]
+            struct Candidate<'tcx> {
+                path: Path<'tcx>,
+                pred: PolyTraitPredicate<'tcx>,
+                origin: AnnotatedPredicate<'tcx>,
+            }
+
+            use std::collections::VecDeque;
+            let mut candidates: VecDeque<Candidate<'tcx>> = starting_points
+                .into_iter()
+                .map(|pred| {
+                    let clause = pred.predicate.as_trait_clause();
+                    clause.map(|clause| Candidate {
+                        path: vec![],
+                        pred: clause,
+                        origin: pred.clone(),
+                    })
+                })
+                .filter(|candidate| candidate.is_some())
+                .map(|candidate| candidate.unwrap())
```

Comment:
Use one [`filter_map`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map) instead of map+filter+map here, we'll get rid of the unwrap :) in general we don't want `unwrap`s at all. Btw, you can use `option.s_unwrap(the_state_of_type_State)` if you really ever need to unwrap: that unwraps with a spanned message.